components/places/doc/sql_concurrency.md

```diff
@@ -0,0 +1,37 @@
+
+Note that multiple writers will only cause `SQLITE_BUSY` after some timeout period has expired. Each writer will block for this timeout period in the hope that the other writer completes and it can begin. The current default is 5 seconds.
+    
+So there are 2 sane options:
+
+* Ensure there's only ever one writer at a time (eg, via a dedicated writer thread) - `SQLITE_BUSY` will never happen, although writing will be queued - a single writer which takes many seconds will block all other writes. This may cause UX issues (eg, visited links not being immediately correct) and given mobile constraints, may never be writen at all (eg, aggressive app termination)
+
+* Allow multiple writers. If you keep the writes short and fast, you should never hit the timeout period and everything should be fine. However, this scenario *does* allow for `SQLITE_BUSY` errors.
+    
+Complicating things is that transactions are not only about integrity, they also improve performance. Consider our history-sync "incoming" implementation - when processing 5k records, we can see a performance improvement of around 5x by using a single transaction for all records vs a transaction per record. We can choose between a fast sync with a long-lived transactions that might exceed our timeout budget, or a slow sync with many short transactions that will be within budget.
+
+But in this scenario there are no "integrity" concerns, just performance. And sadly, it's not really possible to "split" a sync up so that it can be interleaved on a single writer thread - ie, once a sync starts, it needs to complete - and much of the time taken for a sync is on the network - it would probably not be workable for sync to block the only writer thread while it makes network requests.
+
+This leaves us with a dilemma:
+
+* If we use a single writer thread we would use a single-transaction for sync - but even this leaves writes blocked for a very long time (and includes time on the network). ie, writes may have *very* high latency, may never complete if the app is terminated - but would never experience `SQLITE_BUSY`.
+
+* If we allowed multiple writer threads, we would use a transaction-per-record for sync, meaning both threads would almost certainly complete in their timeout budget - although the sync would take many many minutes to complete.
+
+So how about a compromise: we use 2 writer threads, but:
+
+* One is dedicated to the small and fast writes (eg, a visit, creation of a bookmark, etc).
+
+* One is dedicated to Sync, but it uses a strategy whereby we use these "performance transactions", but only for a set period of time. For example, when processing 5k records, we use one transaction per (say) 1000ms. We squeeze as many of the 5k records as we can into this time period and commit and start a new transaction as necessary.
```

Comment:
It's also possible, though maybe inadvisable, for the small writer to preempt the Sync connection on `SQLITE_BUSY`, by calling `sqlite3_interrupt` on the Sync connection. As you pointed out on Slack, it's not immediate, but it will cancel pending reads and roll back transactions...which is fine for Sync! At that point, we know we're syncing stale data, anyway, so we can abort and try the sync again once the main writer finishes.

Reply:
Interesting! One thing that worries me with this is that Sync may end up being interrupted most of the time while the app is being used (ie, that if a user was actually browsing and accumulating visits, sync would always be killed.) This means Sync will continually be doing (possibly significant) work before being killed and needing to do it again.

However, if we could arrange for it to happen only when the small writer would actually get SQLITE_BUSY it might make sense. I wonder if we could abuse sqlite3_busy_handler here?